core/src/test/scala/io/projectglow/gff/GffReaderSuite.scala

```diff
@@ -0,0 +1,280 @@
+  )
+
+  private val testSchemaWithUnderscore = StructType(
+    gffBaseSchema.fields :+
+    StructField("dbx_Ref", ArrayType(StringType))
+    :+ StructField("is_Circular", BooleanType)
+  )
+
+  private val testSchemaExpected = StructType(
+    gffBaseSchema.fields :+
+    StructField("Dbxref", ArrayType(StringType))
+    :+ StructField("Is_circular", BooleanType)
+  )
+
+  test("Schema inference") {
+    val df = spark
+      .read
+      .format(sourceName)
+      .load(testGff3)
+
+    val expectedSchema = StructType(
+      gffBaseSchema.fields.dropRight(1) ++ testOfficialFields ++ testUnofficialFields
+    )
+
+    assert(df.schema.equals(expectedSchema))
+  }
+
+  gridTest("Case-and-underscore-insensitive attribute column names in user-specified schema")(
+    Seq(testSchemaAllSmall, testSchemaMixedCase, testSchemaWithUnderscore)
+  ) { s =>
+    val dfRow = spark
+      .read
+      .format(sourceName)
+      .schema(s)
+      .load(testGff3)
+      .orderBy("seqId", "start")
+      .take(1)(0)
+
+    val expectedRow = Row(
+      "NC_000001.11",
+      "RefSeq",
+      "region",
+      0,
+      248956422,
+      null,
+      "+",
+      1,
+      "ID=NC_000001.11:1..248956422;Dbxref=taxon:9606,test;Name=1;chromosome=1;" +
+      "gbkey=Src;genome=chromosome;mol_type=genomic DNA;Is_circular=false",
+      Array("taxon:9606", "test").toSeq,
+      false
+    )
+    assert(dfRow == expectedRow)
+  }
+
+  gridTest("Read gff3, gzipped gff3 and bgzipped gff3 with inferred schema")(
+    Seq(testGff3, testGff3Gzip, testGff3Bgzip, testGff3BgzipWithGzSuffix)
+  ) { file =>
+    val dfRows = spark
+      .read
+      .format(sourceName)
+      .load(file)
+      .orderBy("seqId", "start", "source")
+      .take(2)
+
+    val expectedRows = Array(
+      Row(
+        "NC_000001.11",
+        "RefSeq",
+        "region",
+        0,
+        248956422,
+        null,
+        "+",
+        1,
+        "NC_000001.11:1..248956422",
+        "1",
+        null,
+        Array("taxon:9606", "test").toSeq,
+        false,
+        "1",
+        null,
+        "Src",
+        null,
+        null,
+        null,
+        "chromosome",
+        "genomic DNA",
+        null,
+        null,
+        null,
+        null
+      ),
+      Row(
+        "NC_000001.11",
+        "BestRefSeq",
+        "pseudogene",
+        11873,
+        14409,
+        null,
+        null,
+        null,
+        "gene-DDX11L1",
+        "DDX11L1",
+        null,
+        Array("GeneID:100287102", "HGNC:HGNC:37102").toSeq,
+        null,
+        null,
+        "DEAD/H-box helicase 11 like 1 (pseudogene)",
+        "Gene",
+        "DDX11L1",
+        "transcribed_pseudogene",
+        null,
+        null,
+        null,
+        null,
+        "true",
+        "passed",
+        null
+      )
+    )
+    assert(dfRows.sameElements(expectedRows))
+  }
+
+  test("Read gff with user-specified schema containing attributesField") {
```

Comment:
What's the expected behavior if the user provides a schema with fields that are not actually in the attributes map? We should have a test for that.

Reply:
The column will be added. All entries will be null. Added a test for that.